### DISCUSSION TITLE
Adding new bonds and angles with custom updater

### DISCUSSION THREAD

**Discussion by nmendiboure** [glotzerlab/hoomd-blue#1867] [Support] [answered]
Hello !

I'm currently trying to create a new bond between 2 beads of 2 different polymers if they are close by a certain distance. I'm using a Custom.Action class and the `snap = self._state.get_snapshot()` function. But when I want to add new bond groups, or new bond typeid, or even angles groups, it simply says:

`AttributeError: property of 'BondDataSnapshot' object has no setter`

As I understand it, I can only modify existing elements, but not add new ones...

Am I missing something that would allow me to define new bonds and angles, or do I need to convert my snapshot to a gsd snapshot and start a new simulation from there? 

```
def update_snapshot_data(snapshot_data, new_groups, new_typeids):

    number_of_entries = len(new_groups)

    groups = snapshot_data.group if snapshot_data.N else []
    groups = list(groups) + list(new_groups)
    snapshot_data.group = np.asarray(groups, dtype=np.uint32)

    typeids = snapshot_data.typeid if snapshot_data.N else []
    typeids = list(typeids) + list(new_typeids)
    snapshot_data.typeid = np.asarray(typeids, dtype=np.uint32)

    snapshot_data.N += number_of_entries

```

Thanks in advance

**Comment by joaander** [comment, marked as answer]
HOOMD's snapshot data structure is not Pythonic - it is a thin wrapper around a C++ data structure. You cannot overwrite the `.group` attribute because it is managed by C++.

To expand the group array, *first* set the new N:
```
snapshot_data.N += number_of_entries
```

Now the `snapshot_data.group` array is larger and you can write to its contents with:
```
snapshot_data.group[:] = np.asarray(groups, dtype=np.uint32)
```

**Reply by nmendiboure** [reply]
Hi thanks for the answer. 

I managed to handle that by converting my `hoomd.Snapshot` object into a `gsd.Frame`. I then modified the gsd file, converted it back into a Snapshot, and applied the changes using `self._state.set_snapshot(snap)`. It seems to work well.

I saw that there was a function to convert a gsd into a Snapshot : `hoomd.Snapshot.from_gsd_frame(gsd, communicator)` , but I didn't find one to do the opposite, so here mine :


```
def snap_to_gsd(snap):
    snap_gsd = gsd.hoomd.Frame()
    for attr in snap_gsd.__dict__:
        data_gsd = getattr(snap_gsd, attr)

        if hasattr(snap, attr):
            data = getattr(snap, attr)

            if hasattr(data_gsd, '__dict__'):
                for prop in data_gsd.__dict__:
                    if hasattr(data, prop):
                        setattr(data_gsd, prop, getattr(data, prop))
                    elif hasattr(data, f"{prop.lstrip('_')}"):
                        # If a private attribute, get it with leading underscore
                        setattr(data_gsd, prop, getattr(data, f"{prop.lstrip('_')}"))

    return snap_gsd
```

**Comment by joaander** [comment]
I would also welcome contributions that add a Python API to create particles/bonds/dihedrals/etc.. See the proposal in #1535.
